pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,341 @@
+		if object.key == "" {
+			continue
+		}
+		if err := f(object.key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, b.name, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, b.name, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+// TODO(bplotka): Consider removing Exists in favor of helper that do Get & IsObjNotFoundErr (less code to maintain).
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	obj, err := b.getRange(ctx, b.name, name, 0, 2)
+	if err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "get bos object metadata")
+	}
+	obj.Close()
+	return true, nil
+}
+
+func (b *Bucket) Close() error {
+	return nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	switch tmpErr := err.(type) {
+	case *bce.Error:
+		if tmpErr.StatusCode == http.StatusNotFound {
+			return true
+		}
+	}
+	return false
+}
+
+func (b *Bucket) getRange(ctx context.Context, bucketName, objectKey string, off, length int64) (io.ReadCloser, error) {
+	if len(objectKey) == 0 {
+		return nil, errors.Errorf("given object name should not empty")
+	}
+
+	getObjectRequest := &bos.GetObjectRequest{
+		BucketName: bucketName,
+		ObjectKey:  objectKey,
+	}
+
+	if length != -1 {
+		if err := setRange(getObjectRequest, off, off+length-1); err != nil {
+			return nil, err
+		}
+	}
+
+	if obj, err := b.client.GetObjectFromRequest(*getObjectRequest, nil); err != nil {
+		return nil, err
+	} else {
+		return obj.ObjectContent, nil
+	}
+}
+
+func setRange(getObjectRequest *bos.GetObjectRequest, start, end int64) error {
+	if start >= 0 && end <= 0 {
+		getObjectRequest.Range = fmt.Sprintf("%v-", start)
+	} else if start >= 0 && start <= end {
+		getObjectRequest.Range = fmt.Sprintf("%v-%v", start, end)
+	} else {
+		return errors.Errorf("Invalid range specified: start=%d end=%d", start, end)
+	}
+	return nil
+}
+
+type objectInfo struct {
+	key string
+	err error
+}
+
+func (b *Bucket) listObjects(ctx context.Context, objectPrefix string) <-chan objectInfo {
+	objectsCh := make(chan objectInfo, 1)
+
+	go func(objectsCh chan<- objectInfo) {
+		defer close(objectsCh)
+		var marker string
+		for {
+			result, err := b.client.ListObjectsFromRequest(bos.ListObjectsRequest{
+				BucketName: b.name,
+				Prefix:     objectPrefix,
+				Delimiter:  dirDelim,
+				Marker:     marker,
+				MaxKeys:    1000,
+			}, nil)
+			if err != nil {
+				select {
+				case objectsCh <- objectInfo{
+					err: err,
+				}:
+				case <-ctx.Done():
```

Comment:
Maybe we don't need select here, we can check `ctx.Err()` to determine the iterator has been canceled.
```
for {
  if ctx.Err() != nil {
    return
  }
}
```

Further information is [here](https://golang.org/pkg/context/#Context).